engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/debug/DebugEvalNode.java

```diff
@@ -22,7 +23,8 @@ public class DebugEvalNode extends Node {
     evalNode.setTailStatus(BaseNode.TailStatus.TAIL_DIRECT);
   }
 
-  Object execute(CallerInfo callerInfo, State state, Object expression) {
+  Object execute(
+      VirtualFrame requestOwnStackFrame, CallerInfo callerInfo, State state, Object expression) {
```

Comment:
This convention of `requestOwnStackFrame` or `giveMeAStackFrame` is completely new and not documented. It took me a while to  see that this is how you guide annotation processor not to generate the `InlinedCallNode`. Wouldn't it make more sense to add a parameter to the `@BuiltinMethod` annotation with sensible defaults and proper documentation?

Reply:
That was my original idea as well and I even prototyped it. But then it turned out `VirtualFrame` works well too and I don't have to change many definitions of `@BuiltinMethod` that are already using it - everything worked out of the box. Then I just needed to fix few tests: c164d95b04aa59edafe56d08606ec1f622ddef28 Only this change is slightly unnatural (but covered with tests) otherwise everything would work out of the box.

I guess we can leave the annotation attribute for later, when we needed.

Reply:
IMO the convention is OK, but indeed it would be good to have it documented somewhere, maybe next to `needsFrame()`?

Reply:
+1 to Hubert's comment. It is very unclear to me that to force a builtin method to be inlined, I need to add `VirtualFrame` parameter to its `execute` method. Moreover, this is counterintuitive to the conventions used in Truffle. Please, introduce some kind of `boolean forceInline` parameter for `@BuiltinMethod` annotation. Not only will it be properly documented, and visible at first sight, but also additional assertions can be added to `MethodProcessor`, e.g., if a builtin method has forceInline == true, then it should have `VirtualFrame` as parameter to its `execute` method.

> I guess we can leave the annotation attribute for later, when we needed.

An refactor issue will be created, and most likely never addressed, only adding more technical debt to already so complex `MethodProcessor`. Anybody fiddling with `MethodProcessor` in the future will spend additional time trying to understand this logic.

Reply:
(Just _note_ that I think the convention is reverse - the method is _inlined_ if it **does not have** a `VirtualFrame` - so _adding_ the `VirtualFrame` argument makes it be **not** inlined.

But I agree 100% with your points on documentation.

Reply:
> (Just note that I think the convention is reverse - the method is inlined if it does not have a VirtualFrame - so adding the VirtualFrame argument makes it be not inlined.

Right, I got that one wrong, sorry.

> [needsFrame()](https://github.com/enso-org/enso/pull/6255/files#diff-3bdc4a1027379741ab7cb736f817ae5ef52687f8ed7a4d4adf84c1c6cf951d61R200) has exactly the same level of documentation as needsCallerInfo() just above. needsCallerInfo() also influences the structure of the generated code. I don't see a reason to treat one or the other in some special way.

But `needsCallerInfo` only influences it negligibly - it only adds one more parameter that is passed to the builtin method's `execute` method. On the other hand, `needsFrame` has a huge influence. Not only to the generated code, but to the whole semantics of the builtin method execution.

Reply:
> counterintuitive to the conventions used in Truffle

Truffle DSL treats `VirtualFrame` in a special way at various places. It is not like the other types, it has a special treatment. It can be omitted (from specializations - unlike other types), sometimes it has to be omitted (when the specialization method has `@TruffleBoundary` annotation), etc.

I use the above examples to demonstrate `VirtualFrame` is _very special_ and its presence/absence _means something_ even in the Truffle DSL.

> introduce some kind of `boolean forceInline` parameter for `@BuiltinMethod` annotation. ... properly documented, and visible at first sight, ... additional assertions can be added to `MethodProcessor`, 

I'd agree, if we were building API for wide use. However `MethodProcessor` & co. is an internal DSL only used in our project...

> technical debt to already so complex `MethodProcessor`.

... by few of us. That doesn't mean we should do poor engineering - I've already felt a lot of pain when dealing with the `MethodProcessor` - increased by the fact that I am not yet sure how to easily debug that code - however I don't share the optimism that adding `allowInline` into `@BuiltinMethod` and `Builtin.Method` and new error messages into `MethodProcessor` is going to decrease the pain.

Alas...
 
> Anybody fiddling with `MethodProcessor` in the future...

...should remember to use `git blame` to find out motivations behind `MethodProcessor` code. It is my hope that lines leading to #6255 pull request will give such curious developers more explanation than (few) hundred lines of additional code to support `allowInline` attribute & co. 

> > leave the annotation attribute for later, when needed.
> An refactor issue will be created, and most likely never addressed, 

I said _"when needed"_ - no refactoring is planned unless there is a performance benefit in doing so. When there was a performance benefit, the issue wouldn't be forgotten.

Reply:
I am aware that my view that [Pull Requests and Issues are part of the project documentation](https://github.com/enso-org/enso/pull/6255#discussion_r1167390077) hasn't been discussed anywhere, but for the sake of ...

> lines leading to #6255 pull request will give ... more explanation

I suggest to discuss it elsewhere than in this PR should such discussion be needed.

Reply:
OK, I started a discussion https://github.com/orgs/enso-org/discussions/6296